File: backend/src/main/java/com/blistra/notifications/domain/Reminder.java
package com.blistra.notifications.domain;

import com.blistra.users.domain.User;
import jakarta.persistence.*;

import java.time.LocalDateTime;
import java.time.OffsetDateTime;
import java.util.UUID;

/**
 * A reminder: "the user wants something to happen at a certain time".
 *
 * <p>Reminders describe <em>when</em> to remind and <em>how</em> to deliver;
 * they do not own the underlying business event. For domain-generated reminders
 * the owning module is authoritative and {@code sourceId} is a soft reference
 * into that module's record (deliberately not a foreign key because the target
 * table differs by {@link ReminderType}).</p>
 *
 * <p>{@code scheduledAt} is an absolute instant stored in UTC (TIMESTAMPTZ).
 * {@code timezone} preserves the user's IANA timezone for display and for
 * client-side local scheduling in the intended local time.</p>
 */
@Entity
@Table(name = "reminders", indexes = {
        @Index(name = "idx_reminders_user_scheduled", columnList = "user_id, scheduled_at"),
        @Index(name = "idx_reminders_user_status", columnList = "user_id, status")
})
public class Reminder {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(columnDefinition = "UUID")
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false, updatable = false)
    private User user;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private ReminderType type;

    @Column(nullable = false, length = 160)
    private String title;

    @Column(length = 500)
    private String body;

    @Column(name = "scheduled_at", nullable = false)
    private OffsetDateTime scheduledAt;

    @Column(nullable = false, length = 64)
    private String timezone;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private ReminderStatus status;

    @Column(name = "source_id", columnDefinition = "UUID")
    private UUID sourceId;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    public Reminder() {
        this.status = ReminderStatus.SCHEDULED;
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
    }

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public ReminderType getType() {
        return type;
    }

    public void setType(ReminderType type) {
        this.type = type;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getBody() {
        return body;
    }

    public void setBody(String body) {
        this.body = body;
    }

    public OffsetDateTime getScheduledAt() {
        return scheduledAt;
    }

    public void setScheduledAt(OffsetDateTime scheduledAt) {
        this.scheduledAt = scheduledAt;
    }

    public String getTimezone() {
        return timezone;
    }

    public void setTimezone(String timezone) {
        this.timezone = timezone;
    }

    public ReminderStatus getStatus() {
        return status;
    }

    public void setStatus(ReminderStatus status) {
        this.status = status;
    }

    public UUID getSourceId() {
        return sourceId;
    }

    public void setSourceId(UUID sourceId) {
        this.sourceId = sourceId;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: frontend/lib/features/notifications/models/reminder.dart
import 'reminder_type.dart';

/// A reminder as returned by the backend REST API.
///
/// `scheduledAt` is an absolute instant kept in UTC (ISO-8601 from the
/// backend's `TIMESTAMPTZ`). `timezone` is the user's IANA zone that was used
/// to choose the wall-clock time, which we keep for display only.
class Reminder {
  Reminder({
    required this.id,
    required this.type,
    required this.title,
    this.body,
    required this.scheduledAt,
    required this.timezone,
    required this.status,
    this.sourceId,
  });

  final String id;
  final ReminderType type;
  final String title;
  final String? body;

  /// The absolute moment this reminder should fire, stored in UTC.
  final DateTime scheduledAt;

  /// IANA timezone id associated with the reminder (display purposes).
  final String timezone;
  final ReminderStatus status;

  /// Soft reference into the owning domain record for non-GENERAL reminders.
  final String? sourceId;

  bool get isActive => status == ReminderStatus.scheduled;

  bool get needsDelivery => isActive;

  factory Reminder.fromJson(Map<String, dynamic> json) {
    return Reminder(
      id: json['id'] as String,
      type: ReminderType.fromWire(json['type'] as String),
      title: json['title'] as String,
      body: json['body'] as String?,
      scheduledAt: DateTime.parse(json['scheduledAt'] as String).toUtc(),
      timezone: json['timezone'] as String,
      status: ReminderStatus.fromWire(json['status'] as String),
      sourceId: json['sourceId'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'type': type.wire,
        'title': title,
        if (body != null) 'body': body,
        'scheduledAt': scheduledAt.toIso8601String(),
        'timezone': timezone,
        'status': status.wire,
        if (sourceId != null) 'sourceId': sourceId,
      };
}